pkg/credentialprovider/azure/azure_credentials.go

```diff
@@ -237,10 +267,10 @@ func (a *acrProvider) Provide(image string) credentialprovider.DockerConfig {
 				cfg[customAcrSuffix] = *cred
 			}
 		}
-
-		// add ACR anonymous repo support: use empty username and password for anonymous access
-		cfg["*.azurecr.*"] = defaultConfigEntry
 	}
+
+	a.expiration = time.Now().Add(a.lifetime)
```

Comment:
This is lengthening the expiration time on existing credentials copied from any previous cacheDockerConfig (line 203). Do the acquired tokens expire? if so, that means requests for images in new registries can extend the cache expiration even as individual items in cacheDockerConfig are no longer valid.

Reply:
Having a single cache expiration and cached credential represent data for distinct registries and with distinct expirations is very hard to follow and error prone. Keying the cache by the registry of the incoming image and managing cached credentials+expirations per-registry would make more sense.

Reply:
thanks @liggitt I have updated this PR using per-registry cache (also with cache expiration), pls note that per-registry cache is only for managed identity, and for service principal(account name & password), it's not necessary since password is supposed to be the same in whole kubelet life time.

Reply:
that looks much better, thanks. one last comment on how we're adding in anonymous credentials to avoid a data race